SDP-47 Adoption of StreamCreate method

igkon opened the pull request:
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

  Add body for submitStreamCreateOperation method which is suitable for new code after refactoring.

Commit:
Adoption of StreamCreate method

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

  Add body for submitStreamCreateOperation method which is suitable for new code after refactoring.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -456,7 +456,46 @@ void submitStreamOperation(final PathOperation streamOp, final OpType opType) {
+			val controller = controllerCache.computeIfAbsent(endpointUri, this::createController);
+			val scopeName = DEFAULT_SCOPE;
+			var streamName = streamOp.item().name();
+			if(streamName.startsWith(SLASH)) {
```
Let's move this duplication code (stream name fixing) into the some method

igkon replied:
Moved

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -456,7 +456,46 @@ void submitStreamOperation(final PathOperation streamOp, final OpType opType) {
+			if(streamName.endsWith(SLASH) && streamName.length() > 1) {
+				streamName = streamName.substring(0, streamName.length() - 1);
+			}
+			val streamConfig = StreamConfiguration
```
stream config should be got from the cache

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -456,7 +456,46 @@ void submitStreamOperation(final PathOperation streamOp, final OpType opType) {
+					.streamName(streamName)
+					.scope(scopeName)
+					.build();
+			if(controller.createStream(streamConfig).get(controlApiTimeoutMillis, TimeUnit.MILLISECONDS)) {
```
we could execute the operation asynchronously if we supply the callback instead
created a ticket for this:
https://mongoose-issues.atlassian.net/browse/PRAVEGA-62

igkon replied:
Ok, then it would be a next PR.

Commit:
Add scope,StreamConfiguration cache.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

  Add cache to hold scopes and StreamConfigurations without stream names. Also move erasing of slash
  into method.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -596,6 +585,16 @@ void closeAllWithTimeout(final Collection<? extends AutoCloseable> closeables) {
 		}
 	}
 
+	private String eraseSlashFromString(String str) {
```
***static*** String extractStreamName(***final*** String itemPath) {

igkon replied:
Done. But do we need to use this method in some other places? Why couldn't it be private?

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -460,19 +457,11 @@ void submitStreamCreateOperation(final PathOperation streamOp, final String node
+			val scopeCreateFunc = scopeCreateFuncCache.computeIfAbsent(controller, ScopeCreateFunctionImpl::new);
+			val streamCreateFunc = streamCreateFuncCache.computeIfAbsent(scopeName, scopeCreateFunc);
+			val streamConfig = scopeStreamConfigsCache.computeIfAbsent(scopeName,streamCreateFunc);
 			var streamName = streamOp.item().name();
```
let's use `val` and pass the `streamOp.item().name()` directly to the function fixing the stream name, without additional mutable local variable

igkon replied:
Ok. It must be better.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -460,19 +457,11 @@ void submitStreamCreateOperation(final PathOperation streamOp, final String node
 			val scopeName = DEFAULT_SCOPE;
+			val scopeCreateFunc = scopeCreateFuncCache.computeIfAbsent(controller, ScopeCreateFunctionImpl::new);
+			val streamCreateFunc = streamCreateFuncCache.computeIfAbsent(scopeName, scopeCreateFunc);
+			val streamConfig = scopeStreamConfigsCache.computeIfAbsent(scopeName,streamCreateFunc);
```
missing whitespace between the args

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -456,7 +453,38 @@ void submitStreamOperation(final PathOperation streamOp, final OpType opType) {
+				completeOperation((O) streamOp, SUCC);
+			} else {
+				Loggers.ERR.debug(
+						"The stream {} already exists in the scope {}", streamName,
```
please fix formatting: single tab indent inside the arguments list, closing bracket on the new line

igkon replied:
Fixed

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -456,7 +453,38 @@ void submitStreamOperation(final PathOperation streamOp, final OpType opType) {
+			val controller = controllerCache.computeIfAbsent(endpointUri, this::createController);
+			val scopeName = DEFAULT_SCOPE;
+			val scopeCreateFunc = scopeCreateFuncCache.computeIfAbsent(controller, ScopeCreateFunctionImpl::new);
+			val streamCreateFunc = streamCreateFuncCache.computeIfAbsent(scopeName, scopeCreateFunc);
```
I think we don't need the stream create function instance, please consider using scope creation function instance to get the corresponding stream config from your cache. You will need your own scope creation function implementation returning the stream configuration instance instead of the stream creation function instance.

Commit:
Fix stream configuration for createStream method

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

  Add new interface and implementation of ScopeCreateFunctionForStreamConfig. It takes scope name, creates scope with that name and returns StreamConfiguration without stream name.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -456,7 +478,46 @@ void submitStreamOperation(final PathOperation streamOp, final OpType opType) {
+			/*We need to use StreamConfiguration.builder() because scopeCreateFuncForStreamConfig returns
+			StreamConfig without streamName field. So this looks like implementation of StreamManager.createStream()
+			 */
+			if(controller.createStream(StreamConfiguration
```
please move the stream config cloning code into some method

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/cache/ScopeCreateFunctionForStreamConfig.java:
```diff
@@ -0,0 +1,19 @@
+	 */
+	@Override
+	StreamConfiguration apply(final String scopeName);
+}
```
please always leave blank line at the end any the source file

Commit:
Add method to combine StreamConfig and stream name

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

  Add method that reconstructs stream configuration with the stream name.

Commit:
Fix extractStreamName method

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>

 Just a little fix after testing mongoose with driver.

Commit:
Fix conflicts with master

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Igor <hokletka@gmail.com>
